src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -594,6 +594,21 @@ public override IEnumerable<T> Enumerable<T>(IQueryExpression queryExpression, Q
 			}
 		}
 
+		public override IAsyncEnumerable<T> AsyncEnumerable<T>(IQueryExpression queryExpression, QueryParameters queryParameters)
+		{
+			using (BeginProcess())
+			{
+				queryParameters.ValidateParameters();
+				var plan = GetHQLQueryPlan(queryExpression, true);
+				AutoFlushIfRequired(plan.QuerySpaces);
+
+				using (SuspendAutoFlush()) //stops flush being called multiple times if this method is recursively called
+				{
+					return plan.PerformAsyncIterate<T>(queryParameters, this);
```

Comment:
Not sure how this will behave....

Reply:
What are you talking about? Is it about  `using (SuspendAutoFlush())`? Yeah this code seems suspicious  - if this suspend is really required  ( as this code looks broken to me even for sync case - it's surely not scoped for items iterations) it needs to be done inside iterator.

Reply:
Yes, I'm talking about that.

Reply:
As now `PerformAsyncIterate` and `PerformIterate` do not execute any query, the auto flush logic was moved into the initialization of the enumerator, so that the auto flush logic will be triggered before the query is executed.